### ISSUE TITLE
char boundary byte indexing panic when using Regex::split

### ISSUE THREAD

**Issue by frewsxcv** [rust-lang/regex#417]
```rust
extern crate regex;

fn main() {
    let a = std::str::from_utf8(b"\\B(?-u)|0").unwrap();
    let b = std::str::from_utf8(b"\n\xcd\x86").unwrap();
    let c = regex::Regex::new(a).unwrap();
    c.split(b).collect::<Vec<_>>(); 
}
```

```
thread 'main' panicked at 'byte index 2 is not a char boundary; it is inside '͆' (bytes 1..3) of `
͆`', src/libcore/str/mod.rs:2232:4
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

[line where the panic happens](https://github.com/rust-lang/regex/blob/d504c82275101d016b125beaf21d64e44bfe099f/src/re_unicode.rs#L834)

found via [afl.rs](https://github.com/rust-fuzz/afl.rs) using [this fuzz target](https://gist.github.com/frewsxcv/7ed273de76a4149ff2632a7a35c0cfd4)

**Comment by BurntSushi** [comment]
I'm away for a bit without access to computer. Does this bug happen in
master?

**Comment by frewsxcv** [comment]
@BurntSushi yep, just confirmed it happens in master too. [here's a backtrace](https://gist.githubusercontent.com/frewsxcv/39c7ec352503b28d356c7940f4e03084/raw/9266986daa182b26a1ff952e3c82e76990ca81b5/gistfile1.txt)

**Comment by BurntSushi** [comment]
This was fixed in `regex 0.2.7`. In particular, negated word boundaries can match invalid UTF-8, which the new regex-syntax crate now detects correctly. Previous it didn't.
